lib/galaxy/tools/__init__.py

```diff
@@ -3282,9 +3283,9 @@ def produce_outputs(self, trans, out_data, output_collections, incoming, history
             passes_filter = in_filter_file if how_filter == "remove_if_absent" else not in_filter_file
 
             if getattr(dce_object, "history_content_type", None) == "dataset":
-                copied_value = dce_object.copy(flush=False)
+                copied_value = dce_object.copy(copy_tags=dce_object.tags, flush=False)
             else:
-                copied_value = dce_object.copy()
+                copied_value = dce_object.copy(copy_tags=dce_object.tags)
```

Comment:
`history_content_type == 'dataset'` covers `HistoryDatasetAssociation` and `LibraryDatasetDatasetAssociation`.
That only leaves`DatasetCollection`, which can't have tag associations.
```suggestion
                copied_value = dce_object.copy()
```